src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -929,125 +725,316 @@ internal ThreadLocalList(int ownerThreadId)
             /// </summary>
             /// <param name="item">The item to add.</param>
             /// <param name="updateCount">Whether to update the count.</param>
-            internal void Add(T item, bool updateCount)
+            internal void LocalPush(T item, bool needsSync)
             {
-                checked
-                {
-                    _count++;
-                }
-                Node node = new Node(item);
-                if (_head == null)
-                {
-                    Debug.Assert(_tail == null);
-                    _head = node;
-                    _tail = node;
-                    _version++; // changing from empty state to non empty state
-                }
-                else
+                Debug.Assert(Environment.CurrentManagedThreadId == _ownerThreadId);
+                bool lockTaken = false;
+                try
                 {
-                    node._next = _head;
-                    _head._prev = node;
-                    _head = node;
+                    // Full fence to ensure subsequent reads don't get reordered before this
+                    Interlocked.Exchange(ref _currentOp, (int)ConcurrentBagListOperation.Add);
+                    int tail = _tailIndex;
+
+                    // We're going to increment the tail; if we'll overflow, then we need to reset our counts
+                    if (tail == int.MaxValue)
+                    {
+                        lock (this)
+                        {
+                            Debug.Assert(_tailIndex == int.MaxValue, "No other thread should be changing _tailIndex");
+
+                            // Rather than resetting to zero, we'll just mask off the bits we don't care about.
+                            // This way we don't need to rearrange the items already in the queue; they'll be found
+                            // correctly exactly where they are.  One subtlety here is that we need to make sure that
+                            // if head is currently < tail, it remains that way.  This happens to just fall out from
+                            // the bit-masking, because we only do this if tail == int.MaxValue, meaning that all
+                            // bits are set, so all of the bits we're keeping will also be set.  Thus it's impossible
+                            // for the head to end up > than the tail, since you can't set any more bits than all of them.
+                            _headIndex = _headIndex & _mask;
+                            _tailIndex = tail = _tailIndex & _mask;
+                            Debug.Assert(_headIndex <= _tailIndex);
+                        }
+                    }
+
+                    // When there are at least 2 elements' worth of space (and there aren't pending operations that
+                    // would force us to synchronize), we can take the fast path without locking.
+                    if (!needsSync && tail < (_headIndex + _mask))
```

Comment:
Ah, no, I see why it's necessary.  Let's say that _array.Length == 32, _headIndex == 0, and _mask == 31.  Let's also say that the array is full, such that _tailIndex == 32.  In a normal case, "_tailIndex <= _headIndex + mask" would return false (32 > 0 + 31) as we'd want.  But let's say that just as we're doing this comparison, a stealing thread comes along and increments _headIndex just as we're doing this.  Then _tailIndex <= _headIndex + mask would return true (32 <= 1 + 31), and we'd end up writing to the element that's currently being stolen (the stealing thread increments the _headIndex before it reads the value).  That's why we need two slots: one to write into, and one to account for the one that may currently be in the process of being stolen.